flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/aggregate/utils/GroupAggHelper.java

```diff
@@ -0,0 +1,156 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.runtime.operators.aggregate.utils;
+
+import org.apache.flink.api.common.state.StateTtlConfig;
+import org.apache.flink.table.data.RowData;
+import org.apache.flink.table.data.utils.JoinedRowData;
+import org.apache.flink.table.runtime.generated.AggsHandleFunction;
+import org.apache.flink.table.runtime.generated.RecordEqualiser;
+import org.apache.flink.table.runtime.operators.aggregate.RecordCounter;
+import org.apache.flink.types.RowKind;
+import org.apache.flink.util.Collector;
+
+import static org.apache.flink.table.data.util.RowDataUtil.isAccumulateMsg;
+import static org.apache.flink.table.data.util.RowDataUtil.isRetractMsg;
+
+/** A helper to do the logic of group agg. */
+public abstract class GroupAggHelper {
+
+    /** Used to count the number of added and retracted input records. */
+    private final RecordCounter recordCounter;
+
+    /** Whether this operator will generate UPDATE_BEFORE messages. */
+    private final boolean generateUpdateBefore;
+
+    /** State idle retention config. */
+    private final StateTtlConfig ttlConfig;
+
+    /** function used to handle all aggregates. */
+    private final AggsHandleFunction function;
```

Comment:
IIUC, this function does not involves state access, right?

Reply:
Yes, currently async state agg does not support accumulate with data view.